### ISSUE TITLE
miniai - Consistency between jumping and incremental selection

### ISSUE THREAD

**Issue by jemag** [echasnovski/mini.nvim#155]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

mini.ai

### Description

Again not quite sure if this qualifies as a feature request, I will try my best to explain the situation.

---

Currently in vim/neovim, consecutive applications of textobjects or textobjects with a count will increment selection. For example,
with cursor on letter `w`, the application of `v3aw` or consecutive applications of `vawawaw` will both go from:

![image](https://user-images.githubusercontent.com/7985687/200660906-f5b7a31f-d7a9-4529-a5fd-d2152336a00c.png)
to
![image](https://user-images.githubusercontent.com/7985687/200660878-67682ca7-fefc-4320-a900-9fd299754769.png)
This is also consistent with other  motions like paragraph, sentence, tags, etc.

However, the behavior does not seem consistent with some of the motions provided by mini.ai. For example, with the argument textobject, pressing `v3aa` or `vaaaaaa` will both go from:

![image](https://user-images.githubusercontent.com/7985687/200661638-0c6ceaf0-7cda-4660-bb80-77a1b9a479a8.png)
to
![image](https://user-images.githubusercontent.com/7985687/200661676-4879154b-ca4e-4a56-91fe-95bd0341f830.png)

Going from vim behavior, expectation would I believe instead be:
![image](https://user-images.githubusercontent.com/7985687/200661902-0698e6c6-47ac-4945-a642-770525960627.png)

This also impacts the "around_next" and "inside_next" text motion where for some situatons the selection will increment and others the selection will jump to the next match.

### Question

Basically, I am wondering if it would be possible to have a relatively consistent approach regarding incremental selection vs jumping. I am not sure what could be the best approach for this:
Perhaps have repeated motions (e.g.: `vawaw` or `vaaaa`) and motions with a count  increment the selection, whereas "around_next" and "inside_next" would always jump to the next occurrence of the textobject?

**Comment by echasnovski** [comment]
The behavior of computing a textobject region is controlled by two configuration settings: `search_method` and `n_lines`. The birds-eye view of the process is the following (see [this help section](https://github.com/echasnovski/mini.nvim/blob/main/doc/mini-ai.txt#L320) for more details):
- Textobject search is done by iterating through all matching regions and picking "the best" one. Which one is considered to be the best is mostly decided by `search_method`. By default it is `cover_or_next` meaning it will try to pick smallest covering region first; if there is no such region, it tries to take closest "next" one. See [here](https://github.com/echasnovski/mini.nvim/blob/07ad48937204446ba350837a3413a5d19de22c36/doc/mini-ai.txt#L457) for more details about search methods.
- Search is done first inside lines of current selection (usually current line) and then in the whole `n_lines` neighborhood (`n_lines` up from top line and same down from bottom line).

> Perhaps have repeated motions (e.g.: vawaw or vaaaa) and motions with a count increment the selection, whereas "around_next" and "inside_next" would always jump to the next occurrence of the textobject?

Coming back to the issue. 'mini.ai' should already do as you suggest. The nuance here is that it does so respecting `search_method`. I believe setting `search_method = 'cover'` will behave more like you assume it should.

> However, the behavior does not seem consistent with some of the motions provided by mini.ai. For example, with the argument textobject, pressing v3aa or vaaaaaa will both go from:

All your examples use argument textobject. What you expect to be an increment here is arguably not an argument, but "two/three arguments". Consecutive application or usage of `count` results into "second/third argument". It is mostly not designed to be incremented, but can result in increment in certain cases (like with the string `f(a, b, c(x, y, z))` and cursor on `x`).

Most other textobjects, though, are designed to be found both as "covering" and "next/prev".

> This also impacts the "around_next" and "inside_next" text motion where for some situatons the selection will increment and others the selection will jump to the next match.

It already should do that. Those are shortcuts for temporary usage of "next" search method. It should select closest matching region which is "to the right" of current reference region (selection) and never result in increasing selection (selecting covering region). If you can provide an example of different behavior, it would be great.

Does this answer your questions?

**Comment by jemag** [comment]
Thank you for the clarification @echasnovski .

It does seem that indeed the behavior is properly consistent with motions implemented by your plugin. 

I guess my confusion stems mostly from the fact of how vim interprets some of its motions compared to how they are interpreted by this plugin, and the fact that you can apply "around_next" and "inside_next" to those vim motions.

For example, if I select a paragraph with `vap` and then use `anp` for around_next paragraph, since those are vim motions it will increment from : 
![image](https://user-images.githubusercontent.com/7985687/200979582-42ad167a-608d-4276-94c0-8151908efafe.png)
to
![image](https://user-images.githubusercontent.com/7985687/200979600-7a25d383-1343-4115-9463-70efd81aff79.png)
So in this case, `vap` followed by `anp` behaves like vim `vapap` and both increment the selection, instead of jumping and selecting the next paragraph which would be the behavior of mini.ai's own motions I believe.

Basically it seems that for vim, it will either keep incrementing (in a lot of textobjects, like word, paragraph, sentence) or in some cases jump (like `)` and `}`). My interpretation is that for textobject that can be consecutive one after another (e.g.: word, paragraph, sentence), it increments, and if not it jumps.

That is why for me I kind of see the `argument` text object similar to word or sentence, as something where you can have multiple consecutive ones in a row, so I guess that is where the incremental selection expectation came from.

--- 

Either way both approaches are probably fine. I think the current approach probably lends itself to greater consistency from within the plugin and easier implementation, although it could diverge slightly from some of vim's original motions.

Hope that long wall of text made sense, please excuse my ramblings.

**Comment by echasnovski** [comment]
> So in this case, vap followed by anp behaves like vim vapap and both increment the selection, instead of jumping and selecting the next paragraph which would be the behavior of mini.ai's own motions I believe.

Sure they are, because `p` is not a 'mini.ai' textobject. It falls back to Neovim's textobject which behaves like you described (in both `ap` and `anp`/`alp`). If implemented as 'mini.ai' custom textobject, it will respect search method (from config in case of `ap`, `next`/`prev` in case of `anp`/`alp`).

> My interpretation is that for textobject that can be consecutive one after another (e.g.: word, paragraph, sentence), it increments, and if not it jumps.

This makes sense in general. Unfortunately, there seems to always be nuances in certain cases, whichever approach to choose.

All in all, I think I'll close this issue, as subject matter seems to be consistent (albeit not always aligned with Neovim's built-ins).
